pkg/daemons/agent/agent_windows.go

```diff
@@ -122,5 +133,77 @@ func kubeletArgs(cfg *config.Agent) map[string]string {
 	if cfg.ProtectKernelDefaults {
 		argsMap["protect-kernel-defaults"] = "true"
 	}
+
 	return argsMap
 }
+
+func waitForSourceVip(networkName string) string {
+	for range time.Tick(time.Second * 5) {
+		network, err := hcsshim.GetHNSNetworkByName(networkName)
+		if err != nil {
+			logrus.WithError(err).Warning("can't find HNS network, retrying", networkName)
+			continue
+		}
+		if network.ManagementIP == "" {
+			continue
+		}
+
+		script := `function GetSourceVip {
+			param(
+					$NetworkName
+				)
+			$hnsNetwork = Get-HnsNetwork | ? Name -EQ $NetworkName.ToLower()
+    		$subnet = $hnsNetwork.Subnets[0].AddressPrefix
+			$ipamConfig = @"
+        {"cniVersion": "0.2.0", "name": "vxlan0", "ipam":{"type":"host-local","ranges":[[{"subnet":"$subnet"}]],"dataDir":"/var/lib/cni/networks"}}
+"@
+			$ipamConfig | Out-File "C:\k\sourceVipRequest.json"
+			$env:CNI_COMMAND="ADD"
+			$env:CNI_CONTAINERID="dummy"
+			$env:CNI_NETNS="dummy"
+			$env:CNI_IFNAME="dummy"
+			$env:CNI_PATH="dummy"
+
+			If(!(Test-Path c:\k\sourceVip.json)){
+				Get-Content c:\k\sourceVipRequest.json | host-local.exe | Out-File c:\k\sourceVip.json
+			}
+
+			$sourceVipJSON = Get-Content c:\k\sourceVip.json | ConvertFrom-Json
```

Comment:
Why are we using powershell to do this? Is this not something that can be done in pure Go?

Reply:
Thanks for such timely review! Just give you the context that I am grabbing the code from how currently the K8s windows worker node being setup, so we know where this integration comes from: 
https://github.com/microsoft/SDN/blob/d1e3be9b6c4fa784764b2e92440a2d364859c417/Kubernetes/windows/kubeadm/KubeCluster.ps1#L269
https://github.com/microsoft/SDN/blob/d1e3be9b6c4fa784764b2e92440a2d364859c417/Kubernetes/windows/helper.v2.psm1#L494

I think we fold the implementation into pure go, but before that, actually I have some questions to this change history:

https://github.com/k3s-io/k3s/commit/483e29e78392b76ba70aa0791120c7a0622ab1eb

So, it seems like sometime before, there was a logic in this file to set up the `source-vip`, but it was removed more recently. Do you know what's the context? Is it just because the code may not work previously? Or for a refactoring reason? Since now I am kind of adding it back, I would like to make sure I am doing it at the correct location. Thanks.

Reply:
See the [PR for that commit](https://github.com/k3s-io/k3s/pull/6517), which notes that it was unused code that was duplicated in RKE2, where it was actually used. If we're going to bring it back, we should probably see if there is an opportunity to make the K3s code usable by both distros.

Reply:
Does it mean maybe make more sense to move the logic into `pkg/daemons/executor/embed.go`? And just want to understand the flow, is the `embed` one being replaced by the RKE2 (pebinaryexecutor?) executor?

Reply:
K3s uses the `embed` executor. RKE2 on Linux uses `podexecutor`, and RKE2 on Windows uses `pebinaryexecutor`. If we need to upstream some or all of the `pebinaryexecutor` from RKE2 here into K3s for k3s on windows, that is probably the correct approach to sharing code.

Reply:
@brandond I made another iteration to move the logic into `embed` executor. Let me know if any further feedback. Thanks!